pkg/controllers/cloud/controller_sync_status.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2023 Antrea Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloud
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"antrea.io/nephe/pkg/logging"
+	"github.com/cenkalti/backoff/v4"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var (
+	ctrlSyncStatus *controllerSyncStatus
+)
+
+type controllerSyncStatus struct {
+	mutex   sync.Mutex
+	syncMap map[controllerType]bool
+	// number of values of the type controllerType
+	numElements int
+	log         func() logging.Logger
+}
+
+func init() {
+	ctrlSyncStatus = &controllerSyncStatus{
+		syncMap:     make(map[controllerType]bool),
+		numElements: int(ControllerTypeVM),
+		log: func() logging.Logger {
+			return logging.GetLogger("controllers-sync-status")
+		},
+	}
+
+	// initialize the map to default values
+	for i := 0; i < ctrlSyncStatus.numElements; i++ {
+		ctrlSyncStatus.syncMap[controllerType(i)] = false
+	}
+}
+
+func GetControllerSyncStatusInstance() *controllerSyncStatus {
+	return ctrlSyncStatus
+}
+
+// SetControllerSyncStatus sets the sync status of the controller.
+func (c *controllerSyncStatus) SetControllerSyncStatus(controller controllerType) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.syncMap[controller] = true
+	c.log().Info("Sync done", "controller", controller.String())
+}
+
+// ResetControllerSyncStatus resets the sync status of the controller.
+func (c *controllerSyncStatus) ResetControllerSyncStatus(controller controllerType) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.syncMap[controller] = false
+}
+
+// IsControllerSynced returns true only when the controller is synced.
+func (c *controllerSyncStatus) IsControllerSynced(controller controllerType) bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.syncMap[controller]
+}
+
+// areControllersSynced returns true only when all the controllers are synced.
+func (c *controllerSyncStatus) areControllersSynced(controllers []controllerType) (bool, error) {
+	for _, name := range controllers {
+		if !c.IsControllerSynced(name) {
+			return false, fmt.Errorf("failed to sync %v controller", name.String())
```

Comment:
something like.. controller %v not synced yet?

Reply:
since its an error message, I want to specify something failed first and then log would explain what failed.
And also to keep the logging consistent, I do the same while waiting for controller to be initialized.

It would look something like this.

`2023-02-14T07:05:06.662Z        ERROR   setup   nephe-controller/main.go:167    problem running manager {"error": "[failed to sync ExternalEntity controller, failed waiting for all runnables to end within grace period of 30s: context deadline exceeded]", "errorCauses": [{"error": "failed to sync ExternalEntity controller"}, {"error": "failed waiting for all runnables to end within grace period of 30s: context deadline exceeded"}]}`

`2023-02-14T07:04:51.554Z        ERROR   setup   controller/controller.go:273    Reconciler error        {"controller": "cloudentityselector", "controllerGroup": "crd.cloud.antrea.io", "controllerKind": "CloudEntitySelector", "cloudEntitySelector": {"name":"cloudentityselector-west-central-us","namespace":"sample-ns"}, "namespace": "sample-ns", "name": "cloudentityselector-west-central-us", "reconcileID": "18f06fd8-b0a2-4fde-9a29-7a4b01e2dbfe", "error": "failed to initialize CloudEntitySelector controller, err: timed out waiting for the condition"}`

Reply:
failed waiting for all runnables to end within grace period of 30s: context deadline exceeded ? 30 sec is the timeout for start?

Reply:
Not really. We don't specify any timeout/deadline in the Start context, this timeout is in the controller runtime.
In nephe case, each controller is waiting for 5 minutes(syncTimeout) in the Start().